onnxruntime/contrib_ops/cuda/bert/attention_softmax.h

```diff
@@ -180,7 +181,16 @@ __device__ inline void SoftmaxWithRawMaskSmall(const int all_sequence_length,
   if (threadIdx.x < all_sequence_length) {
     const int batch_index = blockIdx.y;
     const int sequence_index = blockIdx.x % sequence_length;
-    const int mask_offset = (mask_dimension == 2) ? batch_index * all_sequence_length + threadIdx.x : batch_index * sequence_length * all_sequence_length + sequence_index * all_sequence_length + threadIdx.x;
+    int mask_offset = 0;
+    if (mask_dimension == 2) {
+      mask_offset = batch_index * all_sequence_length + threadIdx.x;
+    } else if (mask_dimension == 3) {
+      mask_offset = batch_index * sequence_length * all_sequence_length + sequence_index * all_sequence_length + threadIdx.x;
+    } else if (mask_dimension == 4){
+      // Megatron code:
+      // ltor_mask = ltor_mask[..., (attention_scores.size(3)-hidden_states.size(1)):attention_scores.size(3), :attention_scores.size(3)]
+      mask_offset = batch_index * max_sequence_length * max_sequence_length + (all_sequence_length - sequence_length) * max_sequence_length + sequence_index * max_sequence_length + threadIdx.x;
+    }
 
     const int& mask = attention_mask[mask_offset];
     float mask_value = mask > 0 ? 0.0f : -10000.0f;
```

Comment:
Need also consider the effect of is_unidirectional and 4d mask as in next few lines.  For example, change 
```
 if (is_unidirectional) {
```
to 
```
 if (is_unidirectional && mask_dimension != 4) {
```